coord/src/rmr/redise.c

```diff
@@ -5,38 +5,188 @@
+      goto error;                                         \
+    }                                                     \
+  })
 
 MRClusterTopology *RedisEnterprise_ParseTopology(RedisModuleCtx *ctx, RedisModuleString **argv,
                                                  int argc) {
+  ArgsCursor ac;
+  ArgsCursor_InitRString(&ac, argv + 1, argc - 1);
+  const char *myID = NULL;                 // Mandatory. No default.
+  size_t numShards = 0;                    // Mandatory. No default.
+  size_t numSlots = 16384;                 // Default.
+  MRHashFunc hashFunc = MRHashFunc_CRC16;  // Default.
 
-  size_t totalLen = 0;
-  const char *cargs[argc];
-  size_t lens[argc];
-  for (int i = 1; i < argc; i++) {
-    cargs[i - 1] = RedisModule_StringPtrLen(argv[i], &lens[i - 1]);
-    totalLen += lens[i - 1] + 1;
+  // Parse general arguments. No allocation is done here, so we can just return on error
+  while (!AC_IsAtEnd(&ac)) {
+    if (AC_AdvanceIfMatch(&ac, "MYID")) {
+      myID = AC_GetStringNC(&ac, NULL);  // Verified after breaking out of loop
+    } else if (AC_AdvanceIfMatch(&ac, "HASHFUNC")) {
+      const char *hashFuncStr = AC_GetStringNC(&ac, NULL);
+      if (!hashFuncStr) {
+        ERROR_MISSING("HASHFUNC");
+        return NULL;
+      }
+      if (!strcasecmp(hashFuncStr, MRHASHFUNC_CRC12_STR)) {
+        hashFunc = MRHashFunc_CRC12;
+      } else if (!strcasecmp(hashFuncStr, MRHASHFUNC_CRC16_STR)) {
+        hashFunc = MRHashFunc_CRC16;
+      } else {
+        ERROR_BADVAL("HASHFUNC", hashFuncStr);
+        return NULL;
+      }
+    } else if (AC_AdvanceIfMatch(&ac, "NUMSLOTS")) {
+      int rc = AC_GetSize(&ac, &numSlots, AC_F_GE1);
+      if (rc != AC_OK) {
+        ERROR_BAD_OR_MISSING("NUMSLOTS", rc);
+        return NULL;
+      } else if (numSlots > 16384) {
+        ERROR_FMT("Bad value for NUMSLOTS: %zu", numSlots);
+        return NULL;
+      }
+    } else if (AC_AdvanceIfMatch(&ac, "RANGES")) {  // End of general arguments
+      int rc = AC_GetSize(&ac, &numShards, AC_F_GE1);
+      if (rc != AC_OK) {
+        ERROR_BAD_OR_MISSING("RANGES", rc);
+        return NULL;
+      }
+      break;
+    } else {
+      ERROR_FMT("Unexpected argument: `%s`", AC_GetStringNC(&ac, NULL));
+      return NULL;
+    }
   }
 
-  char *str = rm_calloc(totalLen + 2, 1);
-  char *p = str;
-  for (int i = 0; i < argc - 1; i++) {
-    strncpy(p, cargs[i], lens[i]);
-    p += lens[i];
-    *p++ = ' ';
-  }
-  p--;
-  *p = 0;
-  RedisModule_Log(ctx, "notice", "Got topology update: %s", str);
-  char *err = NULL;
-  MRClusterTopology *topo = MR_ParseTopologyRequest(str, strlen(str), &err);
-  rm_free(str);
-  if (err != NULL) {
-    RedisModule_Log(ctx, "warning", "Could not parse cluster topology: %s", err);
-    RedisModule_ReplyWithError(ctx, err);
-    rm_free(err);
+  if (!myID) {
+    ERROR_MISSING("MYID");
     return NULL;
   }
 
+  MRClusterTopology *topo = MR_NewTopology(numShards, numSlots, hashFunc);
```

Comment:
Maybe we should check first that `numShards > 0`? I do not think that 0 is a valid number right?

Reply:
We check it using the `AC_F_GE1` flag when parsing the values